_posts/2022-07-20-polymorphic-table-functions.md

```diff
@@ -0,0 +1,193 @@
+Trino is on its way to full support for polymorphic table functions (PTFs).
+
+In this blog post, we are explaining the concept of table functions and 
+exploring how they can be leveraged. We also look at what we have already 
+implemented, and take a sneak peek into the future.
+
+<!--more-->
+
+### Definition time
+
+There are several kinds of functions you can call in a SQL query: scalar
+functions, aggregate functions, and window functions. They might process the
+input row by row (scalar) or all at once (aggregate). One thing they have in
+common is that they return scalar values. Table functions are different. They
+return tables. In a query, they can appear in any place where a table reference
+shows up such as a `FROM` clause:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(my_table_function('foo'));
+```
+
+You can also use table functions in joins:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(my_table_function('bar'))
+JOIN
+  TABLE(another_table_funciton(1, 2, 3))
+ON true;
+```
+
+Polymorphic table functions (PTFs) are a subset of table functions where the
+schema of the returned table is determined dynamically. The returned table
+schema can depend on the arguments you pass to the function.
+
+### OK, but why are we so excited?
+
+We are excited because this feature is a real game changer! Polymorphic table
+functions make SQL extensible and act as a bridge between the Trino engine and
+external systems or resources you might need for processing your data.
+Additionally, polymorphic table functions are standard SQL, and they are very
+convenient to use.
+
+### What is available in Trino today?
+
+So far, we have added a framework for table functions which can be executed by
+the connector. Although this is not the full PTF feature yet, we couldn't wait
+to bring it to life. We added query pass-through table functions for JDBC-based
+connectors and ElasticSearch. They mostly go by the name `query`, and they take
+a single argument, that being the query text:
+
+```sql
+SELECT
+  *
+FROM
+  TABLE(
+    postgresql.system.query(query =>
+        'SELECT
+          name
+        FROM
+          tpch.nation
+        WHERE
+          nationkey = 0'
+    )
+  );
+```
+
+And this will return:
+
+```sql
+  name
+---------
+ ALGERIA
+(1 row)
+```
+
+Something you can’t notice from that example is that when you’re passing that
+“query” argument, it’s pushing the entire query down to PostgreSQL to process
+and run. Whatever connector you’re using, the query argument you pass needs to
+be written so that it works on the underlying database. On the opposite and more
+exciting side of that, if you have a legacy query specific to a database which
+has non-standard SQL syntax and would be difficult to rewrite for Trino, now you
+can pass that entire query down to the connector by wrapping it in the `query`
+function, skipping the need to migrate it.
+
+Besides PostgreSQL, the `query` table function is also available for Druid,
```

Comment:
Technically, these are all different functions that happen to have the same name